src/webgpu/shader/validation/decl/var.spec.ts

```diff
@@ -768,3 +768,84 @@ g.test('var_access_mode_bad_template_delim')
     const ok = t.params.prefix === '<' && t.params.suffix === '>';
     t.expectCompileResult(ok, prog);
   });
+
+g.test('shader_stage')
+  .desc('Test the limitations of address space and shader stage')
+  .params(u =>
+    u
+      .combine('stage', ['compute', 'vertex', 'fragment'] as const)
+      .combine('kind', [
+        'handle_ro',
+        'handle_rw',
+        'function',
+        'private',
+        'storage_ro',
+        'storage_rw',
+        'uniform',
+        'workgroup',
+      ])
+  )
+  .fn(t => {
+    let mdecl = ``;
+    let fdecl = ``;
+    let expect = true;
+    switch (t.params.kind) {
+      case 'handle_ro':
+        mdecl = `@group(0) @binding(0) var v : sampler;`;
+        break;
+      case 'handle_rw':
+        mdecl = `@group(0) @binding(0) var v : texture_storage_2d<r32uint, read_write>;`;
+        expect = t.params.stage !== 'vertex';
+        break;
+      case 'function':
+        fdecl = `var v : u32;`;
+        break;
+      case 'private':
+        mdecl = `var<private> v : i32;`;
+        break;
+      case 'storage_ro':
+        mdecl = `@group(0) @binding(0) var<storage> v : u32;`;
+        break;
+      case 'storage_rw':
+        mdecl = `@group(0) @binding(0) var<storage, read_write> v : u32;`;
```

Comment:
For the tests with `read` and `read_write` should there also be a `write` that always fails? So `storage_w` and `handle_w` tests

Reply:
We don't allow write for storage buffer, but it could be added for handle. In fact, this test needs to skip the handle_rw based on a language feature too!

Reply:
Right, it isn't allowed, but isn't this the place we should have validation that it correctly isn't allowed?

Reply:
That's covered in `address_space_access_mode` test suite.